lib/flight_scheduler/schedulers/backfilling_scheduler.rb

```diff
@@ -29,45 +29,232 @@
+          Async.logger.debug("Potential nodes for reservation") { potential_nodes.map(&:name) }
+          node_availabilities = potential_nodes.map do |node|
+            NodeAvailability.new(node, @candidate)
+          end
+
+          node_availabilities.each do |availability|
+            if !availability.has_guarantee?
+              # There is no guarantee of when this node will be available.
+            elsif availability.jobs_waited_on.empty?
+              grouped[:now] << availability
+            else
+              grouped[availability.available_at] ||= []
+              grouped[availability.available_at] << availability
+            end
+          end
+
+          grouped
+        end
+    end
   end
 
   private
 
   def run_allocation_loop(candidates)
     new_allocations = []
-    backfill = Backfill.new(0, nil)
+    reservations = []
 
     loop do
       candidate = candidates.next
       break if candidate.nil?
       Async.logger.debug("Candidate #{candidate.display_id}")
 
-      if backfill.available && candidate.min_nodes > backfill.available
-        Async.logger.debug("Ignoring candidate. It wants more nodes than can be backfilled")
-        next
-      end
-
-      allocation = allocate_job(candidate)
+      allocation = allocate_job(candidate, reservations: reservations)
       if allocation.nil?
-        calculate_available_backfill(candidate, backfill)
-        Async.logger.debug("Unable to allocate candidate. Backfilling updated") { backfill }
-        if backfill.available > 0
-          next
+        Async.logger.debug("Unable to allocate candidate #{candidate.display_id}. Attempting to backfill.")
+        reservation = ReservationFinder.new(candidate).call
+        if reservation.nil?
+          # There are two possible reasons that the reservation was not
+          # created.
+          #
+          # 1. The partiion does not currently have enough resources to run
+          #    the candidate.  If more resources are powered on, this could
+          #    change.
+          #
+          # 2. The partition is running jobs without a time limit set, which
+          #    prevent the reservation from being created.
+          #
+          # These should be handled differently.
+          #
+          # If there are jobs running on the partition without a time limit,
+          # we abort any further processing of the queue.  This ensure that
+          # `candidate`, which is the highest priority job, will not be
+          # delayed by backfilling.
+          #
+          # If the partition currently has insufficient resources connected,
+          # we skip the candidate and continue with the next normally.  This
+          # ensures that such a job does not block the queue, whilst still
+          # allowing it to eventually run.
+          if job_fits_partition?(candidate)
+            Async.logger.debug("Unable to create reservation. Exiting allocation loop.")
+            break
+          else
```

Comment:
This conditional doesn't check directly if the job has a time-limit, instead it infers the condition from the fit algorithm. This _should_ work fine, however it has the possibility of hiding erroneous scheduling quirks.

Checking if the `candidate` is a fast operation so it might be worth splitting the logging condition. Something like:

```
if job_fit_partition?(candidate)
  ...
elsif candidate.timelimit?
  ...
else
  Async.logger.error "Unexpectedly failed to allocate the highest priority job"
end
```

Reply:
The reservation can fail to be created if the candidate doesn't fit or if there are *other* jobs lacking a time limit.  A time limit on the candidate does not make any difference to whether a reservation can be created or not.  Unfortunately, checking if there are *other* jobs lacking a time limit is not necessarily quick.  You're right that this could hide a bug, but I don't think it will and we have sufficient logging to enable debugging (hopefully).